File: apps/android/app/src/main/java/com/signalasi/chat/AgentConversationAttachmentContinuity.kt
package com.signalasi.chat

import android.content.Context

internal data class AgentPriorAttachmentTurn(
    val turnId: String,
    val blocks: List<AgentRichBlock>
)

internal object AgentConversationAttachmentContinuity {
    fun resolve(
        context: Context,
        conversationId: String,
        currentTurnId: String,
        request: String
    ): List<AgentInputAttachment> {
        val selected = select(
            AgentTranscriptStore(context.applicationContext).list(conversationId),
            currentTurnId,
            request
        ) ?: return emptyList()
        return AgentAttachmentWorkspaceStager.restore(
            context.applicationContext,
            conversationId,
            selected.turnId,
            selected.blocks
        )
    }

    internal fun select(
        entries: List<AgentTranscriptEntry>,
        currentTurnId: String,
        request: String
    ): AgentPriorAttachmentTurn? {
        val candidates = entries.asReversed().mapNotNull { entry ->
            if (
                entry.role != AgentTranscriptRole.USER ||
                entry.turnId.isBlank() ||
                entry.turnId == currentTurnId
            ) return@mapNotNull null
            val blocks = AgentRichContentCodec.decode(entry.richOutputJson).filter { block ->
                block.type in ATTACHMENT_TYPES &&
                    block.metadata["source"].orEmpty() == "user_attachment"
            }
            blocks.takeIf { it.isNotEmpty() }?.let {
                AgentPriorAttachmentTurn(entry.turnId, it)
            }
        }
        if (candidates.isEmpty()) return null
        val currentRequest = currentRequest(request)
        candidates.firstOrNull { candidate ->
            candidate.blocks.any { block ->
                val name = block.title.trim()
                name.isNotBlank() && currentRequest.contains(name, ignoreCase = true)
            }
        }?.let { return it }
        return candidates.firstOrNull()?.takeIf {
            referencesPriorArtifact(currentRequest)
        }
    }

    internal fun referencesPriorArtifact(request: String): Boolean {
        val normalized = request.trim().lowercase().split(Regex("\\s+")).joinToString(" ")
        if (normalized.isBlank()) return false
        return CONTINUATION_TERMS.any { term -> containsTerm(normalized, term) }
    }

    private fun currentRequest(value: String): String {
        val explicit = value.substringAfterLast(CURRENT_REQUEST_MARKER, "")
        if (explicit.isNotBlank()) return explicit.trim()
        val contextEnd = value.lastIndexOf(AgentTranscriptStore.SIGNALASI_CONTEXT_TRANSPORT_FOOTER)
        if (contextEnd >= 0) {
            return value.substring(
                contextEnd + AgentTranscriptStore.SIGNALASI_CONTEXT_TRANSPORT_FOOTER.length
            ).trim()
        }
        return value.trim()
    }

    private fun containsTerm(content: String, term: String): Boolean {
        if (!term.isAsciiWord()) return term in content
        return Regex("(?<![a-z0-9])${Regex.escape(term)}(?![a-z0-9])").containsMatchIn(content)
    }

    private fun String.isAsciiWord(): Boolean = isNotBlank() && all { it in 'a'..'z' }

    private const val CURRENT_REQUEST_MARKER = "Current user request:"
    private val ATTACHMENT_TYPES = setOf(
        AgentRichBlockType.IMAGE,
        AgentRichBlockType.FILE,
        AgentRichBlockType.VIDEO,
        AgentRichBlockType.AUDIO
    )
    private val CONTINUATION_TERMS = listOf(
        "continue", "again", "same", "previous", "above", "this", "that", "it",
        "modify", "edit", "revise", "change", "correct", "fix", "improve",
        "annotate", "mark", "redo", "regenerate", "return", "send", "save",
        "download", "export", "open", "inspect", "analyze", "summarize",
        "translate", "crop", "rotate", "resize", "compress", "every", "each",
        "all", "image", "photo", "file", "document", "attachment", "answer",
        "question",
        "\u7ee7\u7eed", "\u518d", "\u540c\u4e00", "\u4e0a\u4e00",
        "\u4e4b\u524d", "\u521a\u624d", "\u8fd9", "\u90a3",
        "\u4fee\u6539", "\u6539", "\u4fee\u6b63", "\u6279\u6539",
        "\u6807\u6ce8", "\u91cd\u505a", "\u91cd\u65b0", "\u4f18\u5316",
        "\u4ed4\u7ec6", "\u6b63\u786e", "\u6bcf\u4e00", "\u9010\u4e00",
        "\u5168\u90e8", "\u90fd", "\u539f\u56fe", "\u56fe\u7247",
        "\u7167\u7247", "\u6587\u4ef6", "\u6587\u6863", "\u9644\u4ef6",
        "\u7b54\u6848", "\u95ee\u9898", "\u53d1\u56de", "\u4fdd\u5b58",
        "\u4e0b\u8f7d", "\u5bfc\u51fa", "\u6253\u5f00", "\u67e5\u770b",
        "\u5206\u6790", "\u603b\u7ed3", "\u7ffb\u8bd1", "\u88c1\u526a",
        "\u65cb\u8f6c", "\u653e\u5927", "\u7f29\u5c0f", "\u538b\u7f29"
    )
}

File: apps/ios/GalaxySSI/AgentConversationAttachmentContinuity.swift
import Foundation

struct AgentAttachmentContinuityReuse {
  var attachments: [GalaxySSIDraftAttachment]
  var stagedAttachments: [AgentStagedAttachment]
}

struct AgentConversationVisualReference: Equatable {
  var turnId: String
  var blocks: [AgentRichBlock]
}

enum AgentConversationAttachmentContinuity {
  private struct PriorTurn {
    var turnId: String
    var blocks: [AgentRichBlock]
  }

  static func resolve(
    conversationId: String,
    currentTurnId: String,
    request: String,
    messages: [ChatMessage]
  ) -> AgentAttachmentContinuityReuse {
    guard let prior = select(messages: messages, currentTurnId: currentTurnId, request: request) else {
      return AgentAttachmentContinuityReuse(attachments: [], stagedAttachments: [])
    }
    let restored = AgentAttachmentWorkspaceStager.restore(
      conversationId: conversationId,
      turnId: prior.turnId,
      blocks: prior.blocks
    )
    return AgentAttachmentContinuityReuse(
      attachments: restored.map(\.attachment),
      stagedAttachments: restored.map(\.staged)
    )
  }

  static func latestVisualReference(
    messages: [ChatMessage],
    currentTurnId: String
  ) -> AgentConversationVisualReference? {
    messages.reversed().compactMap { message -> AgentConversationVisualReference? in
      guard message.isMine,
            !message.isSystem,
            !message.turnId.isEmpty,
            message.turnId != currentTurnId else { return nil }
      let images = AgentRichContentCodec.decode(message.richOutputJson).filter {
        $0.type == .image && $0.metadata["source"] == "user_attachment"
      }
      return images.isEmpty ? nil : AgentConversationVisualReference(turnId: message.turnId, blocks: images)
    }.first
  }

  private static func select(
    messages: [ChatMessage],
    currentTurnId: String,
    request: String
  ) -> PriorTurn? {
    let candidates = messages.reversed().compactMap { message -> PriorTurn? in
      guard message.isMine,
            !message.isSystem,
            !message.turnId.isEmpty,
            message.turnId != currentTurnId else {
        return nil
      }
      let blocks = AgentRichContentCodec.decode(message.richOutputJson).filter {
        attachmentTypes.contains($0.type) && $0.metadata["source"] == "user_attachment"
      }
      return blocks.isEmpty ? nil : PriorTurn(turnId: message.turnId, blocks: blocks)
    }
    guard !candidates.isEmpty else { return nil }
    let currentRequest = currentRequest(request)
    if let named = candidates.first(where: { candidate in
      candidate.blocks.contains { block in
        let name = block.title.trimmingCharacters(in: .whitespacesAndNewlines)
        return !name.isEmpty && currentRequest.range(of: name, options: [.caseInsensitive, .diacriticInsensitive]) != nil
      }
    }) {
      return named
    }
    if referencesPriorArtifact(currentRequest) {
      return candidates.first
    }
    if let visual = latestVisualReference(messages: messages, currentTurnId: currentTurnId) {
      return PriorTurn(turnId: visual.turnId, blocks: visual.blocks)
    }
    return nil
  }

  private static func currentRequest(_ value: String) -> String {
    guard let marker = value.range(of: "Current user request:", options: [.caseInsensitive, .backwards]) else {
      return value.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    return String(value[marker.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private static func referencesPriorArtifact(_ request: String) -> Bool {
    let normalized = request
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .lowercased()
    guard !normalized.isEmpty else { return false }
    return continuationTerms.contains { term in
      term.allSatisfy({ $0.isASCII && $0.isLetter })
        ? normalized.range(of: "(?<![a-z0-9])\(NSRegularExpression.escapedPattern(for: term))(?![a-z0-9])", options: .regularExpression) != nil
        : normalized.contains(term)
    }
  }

  private static let attachmentTypes: Set<AgentRichBlockType> = [.image, .file, .video, .audio]
  private static let continuationTerms = [
    "continue", "again", "same", "previous", "above", "this", "that", "it",
    "modify", "edit", "revise", "change", "correct", "fix", "improve",
    "annotate", "mark", "redo", "regenerate", "return", "send", "save",
    "download", "export", "open", "inspect", "analyze", "summarize",
    "translate", "crop", "rotate", "resize", "compress", "image", "photo",
    "file", "document", "attachment", "\u{7ee7}\u{7eed}", "\u{518d}",
    "\u{540c}\u{4e00}", "\u{4e0a}\u{4e00}", "\u{4e4b}\u{524d}",
    "\u{4fee}\u{6539}", "\u{6807}\u{6ce8}", "\u{91cd}\u{505a}",
    "\u{4f18}\u{5316}", "\u{56fe}\u{7247}", "\u{7167}\u{7247}",
    "\u{6587}\u{4ef6}", "\u{6587}\u{6863}", "\u{9644}\u{4ef6}",
    "\u{5206}\u{6790}", "\u{603b}\u{7ed3}", "\u{7ffb}\u{8bd1}",
    "\u{88c1}\u{526a}", "\u{65cb}\u{8f6c}", "\u{538b}\u{7f29}"
  ]
}
